mitmproxy/proxy/layers/quic.py

```diff
@@ -0,0 +1,1052 @@
+        else:
+            assert not self._ended_server
+            self._ended_server = True
+
+        # we're done if the stream is ignored
+        if self.flow is None:
+            return
+
+        # set and report the first error
+        if err is not None and self.flow.error is None:
+            self.flow.error = mitm_flow.Error(err)
+            yield TcpErrorHook(self.flow)
+
+        # report error-free endings and always clear the live flag
+        if self._ended_client and self._ended_server:
+            if self.flow.error is None:
+                yield TcpEndHook(self.flow)
+            self.flow.live = False
+
+
+class QuicStreamLayer(layer.Layer):
+    """
+    Layer on top of `ClientQuicLayer` and `ServerQuicLayer`, that simply relays all QUIC streams and datagrams.
+    This layer is chosen by the default NextLayer addon if ALPN yields no known protocol.
+    It uses `UDPFlow` and `TCPFlow` for datagrams and stream respectively, which makes message injection possible.
+    """
+
+    buffer_from_client: list[quic_events.QuicEvent]
+    buffer_from_server: list[quic_events.QuicEvent]
+    flow: udp.UDPFlow | None  # used for datagrams and to signal general connection issues
+    quic_client: QuicConnection | None = None
+    quic_server: QuicConnection | None = None
+    streams_by_flow: dict[tcp.TCPFlow, QuicStream]
+    streams_by_id: dict[int, QuicStream]
+
+    def __init__(self, context: context.Context, ignore: bool = False) -> None:
+        super().__init__(context)
+        self.buffer_from_client = []
+        self.buffer_from_server = []
+        if ignore:
+            self.flow = None
+        else:
+            self.flow = udp.UDPFlow(self.context.client, self.context.server, live=True)
+        self.streams_by_flow = {}
+        self.streams_by_id = {}
+
+    def get_or_create_stream(
+        self, stream_id: int
+    ) -> layer.CommandGenerator[QuicStream]:
+        if stream_id in self.streams_by_id:
+            return self.streams_by_id[stream_id]
+        else:
+            # register the stream and start the flow
+            stream = QuicStream(self.context, stream_id, ignore=self.flow is None)
+            self.streams_by_id[stream.stream_id] = stream
+            if stream.flow is not None:
+                self.streams_by_flow[stream.flow] = stream
+                yield TcpStartHook(stream.flow)
+            return stream
+
+    def handle_quic_event(
+        self,
+        event: quic_events.QuicEvent,
+        from_client: bool,
+    ) -> layer.CommandGenerator[None]:
+        # buffer events if the peer is not ready yet
+        peer_quic = self.quic_server if from_client else self.quic_client
+        if peer_quic is None:
+            (self.buffer_from_client if from_client else self.buffer_from_server).append(event)
+            return
+        peer_connection = self.context.server if from_client else self.context.client
+
+        if isinstance(event, quic_events.DatagramFrameReceived):
```

Comment:
Do I see this correctly that DatagramFrameReceived is completely undocumented in aioquic?